compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1938,10 +1948,48 @@ public void visit(BLangRecordDestructure recordDeStmt) {
             return;
         }
         typeChecker.checkExpr(recordDeStmt.expr, this.env);
+        if (checkValidMappingBindingPatternForRecordVarRef(recordDeStmt.varRef.recordRefFields,
+                recordDeStmt.expr.getBType())) {
+            dlog.error(recordDeStmt.expr.pos, DiagnosticErrorCode.INVALID_OPTIONAL_FIELD_IN_MAPPING_BINDING_PATTERN);
+        }
         checkRecordVarRefEquivalency(recordDeStmt.pos, recordDeStmt.varRef, recordDeStmt.expr.getBType(),
                                      recordDeStmt.expr.pos);
     }
 
+    private boolean checkValidMappingBindingPatternForRecordVarRef(List<BLangRecordVarRefKeyValue> recordRefFields,
+                                                                   BType type) {
+        type = types.getReferredType(type);
+        if (type.tag != TypeTags.RECORD) {
+            return false;
+        }
+        BRecordType recordType = (BRecordType) type;
+        for (BLangRecordVarRefKeyValue lhsField : recordRefFields) {
+            if (recordType.fields.containsKey(lhsField.variableName.value)) {
+                if (Symbols.isOptional(recordType.fields.get(lhsField.variableName.value).symbol)) {
```

Comment:
Also need to disallow when the field corresponds to the rest field. The following doesn't fail.

```ballerina
type Employee record {
    string name;
};

function fn() {
    Employee e = {name: ""};
    Employee {name, id} = e;
}
```

Reply:
Related issue - https://github.com/ballerina-platform/ballerina-lang/issues/33538.

Reply:
Aren't we fixing this?